test_runner/regress/test_ondemand_download.py

```diff
@@ -98,6 +101,9 @@ def test_ondemand_download_large_rel(
     # wait until pageserver receives that data
     wait_for_last_record_lsn(client, tenant_id, timeline_id, current_lsn)
 
+    # stop pg before checkpoint to stop wal generation
+    pg.stop()
+
```

Comment:
* shutdown pg
* wait_for_sk_commit_lsn_to_reach_remote_storage
* assert we have generated a bunch of layer files, e.g., using layermap endpoint
* start pageserver (not sks, not postgres)
* bring up read-only postgres (lsn=...) 
   * see https://github.com/neondatabase/neon/blob/a64dd3ecb58f02826d159dfe07a24b7ee52c9b82/test_runner/regress/test_disk_usage_eviction.py#L87-L95 for example
* do the query on the readonly postgres

absence of SKs means we don't have a logical size calculation from walreceiver.

also assert that the log doesn't contain any `logical_size.*download` messages ?

Reply:
Sorry for missing this.

The problem with plain `pg.stop()` (now `endpoint.stop()`) before the subsequent checkpoint is that the data written to pg might not yet have arrived at the PS when taking the checkpoint.

I think my suggestion to use `wait_for_sk_commit_lsn_to_reach_remote_storage` is obsolete, and the `wait_for_last_record_lsn` + moved `endpointstop()` is fine to ensure that the data written to PG reaches the pageserver before checkpoint.

What's still problematic is that the SKs are still running when we restart the pageserver.
That's what I was (badly) pointing out in
> absence of SKs means we don't have a logical size calculation from walreceiver.
> also assert that the log doesn't contain any logical_size.*download messages ?

Basically, if the SKs are still running upon PS restart, the PS is going to connect to the SK and it will calculate logical size for walreceiver status.
That means by the time we

```
before_downloads = get_num_downloaded_layers(client, tenant_id, timeline_id)
```

the on-demand downloads may have already happened and we'll trip

```
after_downloads = get_num_downloaded_layers(client, tenant_id, timeline_id)
log.info(f"layers downloaded before {before_downloads} and after {after_downloads}")
assert after_downloads > before_downloads
```

Hope this helps.

Reply:
I'll put squash auto-merge optimistically, don't push resolve until you are happy with this.